cloudinit/net/netplan.py

```diff
@@ -235,6 +237,61 @@ def _clean_default(target=None):
         os.unlink(f)
 
 
+def netplan_api_write_yaml_file(net_config_content: str) -> bool:
+    """Use netplan.State._write_yaml_file to write netplan config
+
+    Where netplan python API exists, prefer to use of the private
+    _write_yaml_file to ensure proper permissions and file locations
+    are chosen by the netplan python bindings in the environment.
+
+    By calling the netplan API, allow netplan versions to change behavior
+    related to file permissions and treatment of sensitive configuration
+    under the API call to _write_yaml_file.
+
+    In future netplan releases, security-sensitive config may be written to
+    separate file or directory paths than world-readable configuration parts.
+    """
+    try:
+        from netplan.parser import Parser
+        from netplan.state import State
+    except ImportError:
+        LOG.debug(
+            "No netplan python module. Fallback to write %s",
+            CLOUDINIT_NETPLAN_FILE,
+        )
+        return False
+    cfg_dir = mkdtemp()
+    try:
+        with open(os.path.join(cfg_dir, "ci-net-cfg.yaml"), "wb+") as f:
+            f.write(util.encode_text(net_config_content))
+            f.flush()
+            f.seek(0, io.SEEK_SET)
+            parser = Parser()
+            parser.load_yaml(f)
+            state_output_file = State()
+            state_output_file.import_parser_results(parser)
+
+            # Write our desired basename 50-cloud-init.yaml, allow netplan to
+            # determine default root-dir /etc/netplan and/or specialized
+            # filenames or read permissions based on whether this config
+            # contains secrets.
+            state_output_file._write_yaml_file(
```

Comment:
@slyon points here for awareness:
 1. cloud-init is calling this private method from netplan in absence of a public netplan API intended to let netplan choose ultimately what file to write and in which directories. Can we consider this a `safe` method to call that won't change until cloud-init is able to call a public method in the future?
 2. If netplan grows the ability in the future, do you expect the functionality to separate sensitive config form world-readable config will live under this method call?
 3.

Reply:
1. Yes
2. Yes

There is an idea of introducing a public `.write_yaml()` method. But we have to work out some kinks first. `_write_yaml_file()` is good for now, as its underlying CAPI is a public symbol `netplan_state_write_yaml_file()`. And it's also what is being used in the Netplan CLI.

Reply:
@slyon excellent and thank you. I was wondering what you though about a feature request for an optional `header` param to _write_yaml_file in the future that would allow cloud-init to inject custom comments into the file being written?

Reply:
I was already pondering about ways to keep the comments, but that's not currently possible with the underlying `libyaml`. I like the idea of a `header` parameter.

Reply:
Another thing that might be helpful is returning the path(s) of any files written so that at the very least we could add our own header after the fact.